### PR TITLE
Improve parsing of generic declarations

### LINKED ISSUE #211: support default type arguments RFC
Amended RFC here:
https://github.com/Roblox/luau/pull/238#issue-1062352225

some examples from the public RFC and derived from the internal code the new language feature is meant to help with:
```lua
type C<T, U... = (string)>    -- ok, type pack with one element
type D<T, U... = ()>
type E<T..., U... = T...>
type F<T..., U... = ...string>

local function watchQuery<T = any, TVariables = OperationVariables>(
    options: WatchQueryOptions<TVariables>
  ): ObservableQuery<T, TVariables>
end

export type ToxyReducersMap<T = { [string]: any }> = {
  [string]: ToxyReducer<T>
}

export type PolloCache<TSerialized = string> & DataProxy {
  read: <T, TVariables = any>(
    query: HeadOptions<TVariables>,
  ): T | TSerialized
}

export type FieldReadFunction<TExisting = any, TReadResult = TExisting> = (
  existing: SafeReadonly<TExisting> | nil
  options: FieldFunctionOptions,
) -> TReadResult | nil
```

### PR DESCRIPTION
**Pull request by JohnnyMorganz**
- Adds support for default types specified in a generic declaration (`type Foo<X = string> = X`).
Ensures that if one default type has been specified, all further parameters must have a default type
- Fixed parsing to ensure generic types go before generic type packs

**BREAKING CHANGE:** converted `GenericDeclarationParameter` into a struct to allow for default types - the old `GenericDeclarationParameter` enum is now `GenericParameterInfo`

Closes #211

### CODE REVIEW

**Review by matthargett — APPROVED** [review]
great job capturing the cases that shouldn't parse!

File: full-moon/tests/roblox_cases/pass/types_generic_declaration/source.lua [hunk 1]
```diff
@@ -0,0 +1,4 @@
+type Foo<T, U> = (T...) -> U...
```
Comment (matthargett) [inline full-moon/tests/roblox_cases/pass/types_generic_declaration/source.lua:1]: a minor suggestion, make these files for OSS Luau can correctly pass analysis on these files. at some point, we may want to include that cross-check in CI.
```suggestion
--!strict
type Foo<T, U> = (T...) -> U...
```

**Comment by Kampfkarren** [comment]
CI is failing? And not just clippy

**Comment by JohnnyMorganz** [comment]
> CI is failing? And not just clippy

Ah whoops, I accepted matts comment addition but forgot that I need to recreate the snapshots - I'll do that later today

**Pull request merged**